core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/CostBasedScheduleGenerator.scala

```diff
@@ -82,71 +90,187 @@ class CostBasedScheduleGenerator(
   }
 
   /**
-    * Create regions based on only pipelined edges. This does not add the region links.
+    * Partitions a physical plan into Regions and assigns storage URIs in two passes.
+    *
+    * <p><strong>Overview</strong></p>
+    * <ol>
+    *   <li><strong>Region construction:</strong>
+    *     Remove all materialized edges from the DAG and compute undirected connected
+    *     components. The resulting “Region Graph” may contain directed cycles.</li>
+    *   <li><strong>Pass 1 – Output URIs:</strong>
+    *     For each Region, allocate storage URIs on every output port of materialized edges.</li>
+    *   <li><strong>Pass 2 – Input URIs:</strong>
+    *     Re-traverse the same Regions and attach reader URIs on input ports using
+    *     the URIs created in Pass 1.</li>
+    * </ol>
     *
-    * @param physicalPlan The original physical plan without materializations added yet.
-    * @param matEdges     Set of edges to materialize (including the original blocking edges).
-    * @return A set of regions.
+    * <p><strong>Why two passes?</strong></p>
+    * <ul>
+    *   <li>Potential directed cycles in the Region Graph makes a topological
+    *       traversal of regions inpossible.</li>
+    *   <li>To ensure every output URI exists before its corresponding reader is assigned,
+    *       and avoiding “reader before writer” holes, two passes are required.</li>
+    * </ul>
+    *
+    * @param physicalPlan the original physical plan (without materializations)
+    * @param matEdges     edges to be materialized (including blocking edges)
+    * @return a set of [[Region]]s whose [[ResourceConfig]] contains only [[URI]]s for [[PortConfig]]s
+    *         ([[edu.uci.ics.amber.engine.architecture.sendsemantics.partitionings.Partitioning]]
+    *         to be assigned later in
+    *         [[edu.uci.ics.amber.engine.architecture.scheduling.resourcePolicies.ResourceAllocator]])
+    * @see [[edu.uci.ics.amber.engine.architecture.scheduling.config.IntermediateInputPortConfig]]
     */
```

Comment:
I suggest removing those links as they can become dangling easily (adding maintenance effort). developers should be able to find definition of a proper class name in IDE without references.